### PR TITLE
Add the "experimental" retrieve by case Id

### PR DESCRIPTION
**Pull request by KevinJCross**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -10,7 +10,7 @@ plugins {
 }
 
 group 'uk.gov.hmcts.reform'
-version '4.3.1'
+version '4.4.1'
```
Comment (timja) [inline build.gradle]: 4.4.0?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java [hunk 2]
```diff
@@ -201,6 +201,20 @@ CaseDetails submitEventForCitizen(
...
+    /**
+     * Retrieve case by Case Id
+     */
+    @RequestMapping(
+            method = RequestMethod.GET,
+            value = "/cases/{cid}",
+            headers = CONTENT_TYPE + "=" + APPLICATION_JSON_VALUE
+    )
+    CaseDetails getCase(
+            @RequestHeader(AUTHORIZATION) String authorisation,
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: have you tested that these headers are required? Wasn't able to see it in the API but they should be required

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java [hunk 3]
```diff
@@ -201,6 +201,20 @@ CaseDetails submitEventForCitizen(
...
 
+    /**
+     * Retrieve case by Case Id
+     */
+    @RequestMapping(
+            method = RequestMethod.GET,
+            value = "/cases/{cid}",
+            headers = CONTENT_TYPE + "=" + APPLICATION_JSON_VALUE
+    )
+    CaseDetails getCase(
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: story does say to call it `read` ;)

File: build.gradle (hunk 1 above)
Comment (KevinJCross) [reply build.gradle]: new feature rather than a patch fix ?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 2 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: Umm just included them since headers are ignored if not needed.

File: build.gradle (hunk 1 above)
Comment (timja) [reply build.gradle]: it's a new feature so its .4 but you should reset the patch number to 0

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 2 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: we should test it first without to see what happens, have you tested the api with curl / this client?

Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: ill try without ...

Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: they are required ... probably got a request interceptor configured in spring 
I dont like the way you can hide that stuff away in spring ☹️

Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: you can do that in all frameworks, but normally you would inject it into the method so that you could use it... but they are aren't using it

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java [hunk 4]
```diff
@@ -201,6 +201,23 @@ CaseDetails submitEventForCitizen(
             @RequestParam Map<String, String> searchCriteria
     );
 
+    /**
+     * Retrieve case by Case Id
+     */
+    @RequestMapping(
```
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: Since this is `V2` I think we should create a new Client for V2 api's
CoreCaseDataApiV2.java
CcdApiV2.java ?
CoreCaseDataApi.V2.java

This way untangle the Models and add the linking functionality to them

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java [hunk 5]
```diff
@@ -13,19 +14,24 @@
...
     @JsonProperty("case_type_id")
+    @JsonAlias("case_type")
     private String caseTypeId;
 
     @JsonProperty("created_date")
+    @JsonAlias("created_on")
     private LocalDateTime createdDate;
 
-    @JsonProperty("last_modified")
+    @JsonProperty("last_modified_on")
```
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: oops ... we need both

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 4 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: There shouldn't be versioning at that level imo, we're deprecating the old api in this PR and saying use the new one, later we'll remove the old methods and eventually this class will only have the new apis

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java [hunk 6]
```diff
@@ -13,19 +14,25 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class CaseDetails {
 
+    @JsonProperty("id")
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: why is this added?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java [hunk 7]
```diff
@@ -13,19 +14,25 @@
...
     private Long id;
 
+    @JsonProperty("jurisdiction")
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: why is this added?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java [hunk 8]
```diff
@@ -13,19 +14,25 @@
...
 
     @JsonProperty("created_date")
+    @JsonAlias("created_on")
     private LocalDateTime createdDate;
 
     @JsonProperty("last_modified")
+    @JsonAlias("last_modified_on")
     private LocalDateTime lastModified;
 
+    @JsonProperty("state")
```
Comment (timja) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: why is this added?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 6 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: If we are adding properties that are not the names to be consistent we should add them all.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 7 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: does not hurt .. as in consistent with the other properties.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 8 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: same consistency

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 3 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: does what it says on the tin.

I personally think the api clients should  be split between types.
caseWorkerApi
citizenApi 
Makes it easy to see whats meant for what.
getCase really just does what it says on the tin.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 2 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: well the more functional ones are declarative in the routing so there are less surprises.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 4 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: not sure about that ... its a new api and model so we should think about a whole break.
I.e. reference the deprecation to the new api.
Under the hood *I think* they will be using the same connection pool so it's not a big overhead.

Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: Or we could put it on a new interface that this one extends kinda an inheritance/composition ?

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 7 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: its unnecessary, it hurts imo adds more cruft than is needed, json property is only needed when remapping properties, please remove

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 6 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: it makes it harder to tell whats changed though if you are adding everything, adding only the things that you have a reason for makes the code easier to read and easier to tell what is being remapped

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 3 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: personally there shouldn't be 2 different api's are they are identical ;)

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java (hunk 4 above)
Comment (timja) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: i think the complexity and pain for everyone are bigger especially then there's not a new api replacement for everything now, 
if it was a clean upgrade sure, but currently you would have to interchange quite a lot...

Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: We cant compose since feign does not support inheritance 😢

File: src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java (hunk 6 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/ccd/client/model/CaseDetails.java]: when it comes to using lombok it can copy the annotations across to the constructor so you can use @Value  but I dont think it likes the @Alias annotations.

File: src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java [hunk 9]
```diff
@@ -36,10 +40,10 @@
     StartEventResponse startForCaseworker(
             @RequestHeader(AUTHORIZATION) String authorisation,
             @RequestHeader(SERVICE_AUTHORIZATION) String serviceAuthorization,
-            @PathVariable String userId,
-            @PathVariable String jurisdictionId,
-            @PathVariable String caseType,
-            @PathVariable String eventId
+            @PathVariable("userId") String userId,
```
Comment (KevinJCross) [inline src/main/java/uk/gov/hmcts/reform/ccd/client/CoreCaseDataApi.java]: for some reason I had to add these to enable the IT test to pass

File: src/test/java/uk/gov/hmcts/reform/ccd/client/CcdClientTest.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        wireMockServer.shutdownServer();
+    }
+
+    @SneakyThrows
+    private static String loadFile(String filename) {
+        return Resources.toString(ClassLoader.getSystemClassLoader().getResource(filename), UTF_8);
+    }
+
+    @Test
+    @DisplayName("Should be able to call the new Api")
```
Comment (timja) [inline src/test/java/uk/gov/hmcts/reform/ccd/client/CcdClientTest.java]: can you remove new? not really that useful when someone looks at it later
```suggestion
    @DisplayName("Should be able to retrieve a case")
```

Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/ccd/client/CcdClientTest.java]: sure ... was not thinking forward when I wrote it.
